### PR TITLE
HDDS-10965. Add Top N Datanode Latency Metrics for Block Operations

### PR DESCRIPTION
**Pull request by xichen01**
## What changes were proposed in this pull request?
Add the datanode dimension for the `GetBlock`, `PutBlock`, `WriteChunk` and `ReadChunk`, Since the Since a cluster has a very large number of Datanodes, if you export metrics for all Datanodes, this will generate a very large number of metrics, so this adds a `TopNMetrics` type, which will only display information about the Datanodes with the highest latency.

With this Metrics, we should be able to quickly find out if a Datanode is failing (for read operations, we can confirm directly with the Metrics `Tag: name`, for write operations, we need to combine it with the `Pipeline` information, in the Metrics for write operations, `Tag: name` is the Datanode UUID of the Pipeline's leader).

## Example
Configuration
```xml
  <property>
    <name>ozone.xceiver.client.top.metrics.latency.record.threshold.ms.key</name>
    <value>1, 10</value>
  </property>
  <property>
    <name>ozone.xceiver.client.top.metrics.latency.record.count.key</name>
    <value>5</value>
  </property>
```

Metrics

```bash
[root@VM-7-9-centos ~]$ curl -s http://127.0.0.1:9878/prom | grep -v '#' | grep -i top | grep top
xceiver_client_metrics_put_block_exceed10_ms_count_top_n{top="1",name="2384b4cc-6808-489c-a14e-0781a6cbeb5d",hostname="VM-7-9-centos"} 1
xceiver_client_metrics_put_block_exceed1_ms_count_top_n{top="1",name="2384b4cc-6808-489c-a14e-0781a6cbeb5d",hostname="VM-7-9-centos"} 13
xceiver_client_metrics_put_block_exceed1_ms_count_top_n{top="2",name="99ce0d5a-0c29-49b5-94b5-cae68f8bdc39",hostname="VM-7-9-centos"} 9
xceiver_client_metrics_write_chunk_exceed10_ms_count_top_n{top="1",name="2384b4cc-6808-489c-a14e-0781a6cbeb5d",hostname="VM-7-9-centos"} 1
xceiver_client_metrics_write_chunk_exceed1_ms_count_top_n{top="1",name="2384b4cc-6808-489c-a14e-0781a6cbeb5d",hostname="VM-7-9-centos"} 13
xceiver_client_metrics_write_chunk_exceed1_ms_count_top_n{top="2",name="99ce0d5a-0c29-49b5-94b5-cae68f8bdc39",hostname="VM-7-9-centos"} 9
```

## How was this patch tested?
existing test

### CODE REVIEW

**Comment by nandakumar131** [comment]
@xichen01 thanks for working on the change.

Please resolve the merge conflict. Can you also introduce additional unit tests to cover `TopNMetrics`?

**Comment by adoroszlai** [comment]
@xichen01 also, can you please fix the PR title?  I don't think this is specific to S3 Gateway.

**Comment by xichen01** [comment]
> @xichen01 thanks for working on the change.
> 
> Please resolve the merge conflict. Can you also introduce additional unit tests to cover `TopNMetrics`?

OK, I will

**Comment by xichen01** [comment]
> @xichen01 also, can you please fix the PR title? I don't think this is specific to S3 Gateway.

Updated

**Comment by adoroszlai** [comment]
> > can you please fix the PR title? I don't think this is specific to S3 Gateway.
> 
> Updated

Thanks a lot, the new one is much better for clarity.

**Comment by kerneltime** [comment]
Why do we need to filter the metrics for top 10 here and add a lot of processing in the data path? Grafana can sort and showcase top 10. Why don't we just add the datanode label for each latency measured from the client. This way, we can compare client side view of latency on a per datanode basis even if they are below a certain threshold? It might be over all simpler to capture all the data in the client and let grafana sort through and present what we want instead of adding the threshold logic here.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientMetrics.java [hunk 1]
```diff
@@ -97,10 +131,26 @@ public void decrPendingContainerOpsMetrics(ContainerProtos.Type type) {
...
   public void addContainerOpsLatency(ContainerProtos.Type type,
-      long latencyMillis) {
+      long latencyMillis, String datanode) {
+    recordLatencyMetricsIfNeeded(type, latencyMillis, datanode);
     containerOpsLatency.get(type).add(latencyMillis);
   }
 
+  private void recordLatencyMetricsIfNeeded(ContainerProtos.Type type, long latencyMillis,
+      String datanode) {
+    if (latencyMsThresholdsOrder.length > 0 && topLatencyMetricsSet.contains(type)) {
```
Comment (kerneltime) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientMetrics.java:141]: This seems like a lot of computation added in the critical data path.

Comment (xichen01) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientMetrics.java:141]: I think we can do a benchmark, and get a CPU Flame Graphs for a Datanode to check whether it is a heavier operation

Comment (xichen01) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientMetrics.java:141]: I did a benchmark use command `ozone freon ommg --operation READ_KEY -n 1000000 -t 400  --bucket bucket1 --volume s3v --size 4k  --duration=1000`, the command generated about 25000 QPS read.

The `recordLatencyMetricsIfNeeded` only 0.07% on the CPU time slice
image
And the Flame Graphs have been put to this PR's jira ticket https://issues.apache.org/jira/secure/attachment/13069380/client-readkey-25000-4k-cpu.html

**Comment by adoroszlai** [comment]
@kerneltime can you please check if the [latest comment](https://github.com/apache/ozone/pull/6776#issuecomment-2154543779) answers your question?

**Pull request closed without merging**